autoflow/autoflow/parser.py

```diff
@@ -0,0 +1,450 @@
+        # Parameters workflow will receive
+        provided_parameter_names = set(data.get("parameters", {}).keys()).union(
+            self._automatic_parameters
+        )
+        # Required parameters that are not provided
+        missing_parameters = required_parameter_names.difference(
+            provided_parameter_names
+        )
+        if missing_parameters:
+            errors["parameters"] = errors.get("parameters", []) + [
+                f"Missing required parameters {missing_parameters} for workflow '{data['workflow_name']}'."
+            ]
+        # Extra parameters that the workflow is not expecting
+        unexpected_parameters = provided_parameter_names.difference(parameter_names)
+        if unexpected_parameters:
+            errors["parameters"] = errors.get("parameters", []) + [
+                f"Unexpected parameters provided for workflow '{data['workflow_name']}': {unexpected_parameters}."
+            ]
+        if errors:
+            raise ValidationError(errors)
+
+    @post_load
+    def make_workflow_config(self, data, **kwargs) -> WorkflowConfig:
+        """
+        Return the provided workflow config parameters in a WorkflowConfig namedtuple.
+        """
+        return WorkflowConfig(**data)
+
+
+class AvailableDatesSensorSchema(Schema):
+    """
+    Schema for configuration parameters for the available dates sensor.
+
+    Fields
+    ------
+    schedule : str or None
+        Cron string describing the schedule on which the sensor should check
+        available dates, or None for no schedule.
+    cdr_types : list of str, optional
+        Subset of CDR types for which available dates should be found.
+        If not provided, defaults to None.
+    workflows : list of Nested(WorkflowConfigSchema)
+        List of workflows (and associated parameters) that the sensor should run.
+    """
+
+    schedule = ScheduleField(required=True, allow_none=True)
+    cdr_types = fields.List(
+        fields.String(validate=validate.OneOf(["calls", "sms", "mds", "topups"])),
+        missing=None,
+    )
+    workflows = fields.List(fields.Nested(WorkflowConfigSchema), required=True)
+
+
+def parse_workflows_yaml(
+    filename: str, inputs_dir: str
+) -> Tuple[
+    "prefect.environments.storage.Storage",
+    Dict[str, Union["prefect.schedules.Schedule", List[str], List[WorkflowConfig]]],
+]:
+    """
+    Construct workflows defined in an input file.
+
+    Parameters
+    ----------
+    filename : str
+        Name of yaml input file
+    inputs_dir : Path
+        Directory in which input files should be found
+    
+    Returns
+    -------
+    workflows : list of Flow
+        List of prefect workflows
+    run_parameters : dict
+        mapping from workflow names to a list of dicts of parameters for which the workflow should be run
+    """
+    with open(Path(inputs_dir) / filename, "r") as f:
+        workflows_yaml = yaml.safe_load(f)
```

Comment:
Kind of a drag that two config formats are used for the one task?

Reply:
This and the Prefect `config.toml`, you mean? I think the Prefect config setup needs re-working a bit anway - `config.toml` doesn't allow for default settings if an environment variable is missing, and can't use Docker secrets, so I'd like to find a better solution there.